src/mito2/src/sst/index/bloom_filter/creator.rs

```diff
@@ -0,0 +1,530 @@
+
+    /// The provider for intermediate files.
+    temp_file_provider: Arc<TempFileProvider>,
+
+    /// Codec for decoding primary keys.
+    codec: IndexValuesCodec,
+
+    /// Whether the indexing process has been aborted.
+    aborted: bool,
+
+    /// The statistics of the indexer.
+    stats: Statistics,
+
+    /// The global memory usage.
+    global_memory_usage: Arc<AtomicUsize>,
+}
+
+impl BloomFilterIndexer {
+    /// Creates a new bloom filter indexer.
+    pub fn new(
+        sst_file_id: FileId,
+        metadata: &RegionMetadataRef,
+        intermediate_manager: IntermediateManager,
+        memory_usage_threshold: Option<usize>,
+    ) -> Result<Option<Self>> {
+        let mut creators = HashMap::new();
+
+        let temp_file_provider = Arc::new(TempFileProvider::new(
+            IntermediateLocation::new(&metadata.region_id, &sst_file_id),
+            intermediate_manager,
+        ));
+        let global_memory_usage = Arc::new(AtomicUsize::new(0));
+
+        for column in &metadata.column_metadatas {
+            let options =
+                column
+                    .column_schema
+                    .skipping_index_options()
+                    .context(IndexOptionsSnafu {
+                        column_name: &column.column_schema.name,
+                    })?;
+
+            let options = match options {
+                Some(options) if options.index_type == SkippingIndexType::BloomFilter => options,
+                _ => continue,
+            };
+
+            let creator = BloomFilterCreator::new(
+                options.granularity as _,
+                temp_file_provider.clone(),
+                global_memory_usage.clone(),
+                memory_usage_threshold,
+            );
+            creators.insert(column.column_id, creator);
+        }
+
+        if creators.is_empty() {
+            return Ok(None);
+        }
+
+        let codec = IndexValuesCodec::from_tag_columns(metadata.primary_key_columns());
+        let indexer = Self {
+            creators,
+            temp_file_provider,
+            codec,
+            aborted: false,
+            stats: Statistics::new(TYPE_BLOOM_FILTER),
+            global_memory_usage,
+        };
+        Ok(Some(indexer))
+    }
+
+    /// Updates index with a batch of rows.
+    /// Garbage will be cleaned up if failed to update.
+    ///
+    /// TODO(zhongzc): duplicate with `mito2::sst::index::inverted_index::creator::InvertedIndexCreator`
+    pub async fn update(&mut self, batch: &Batch) -> Result<()> {
+        ensure!(!self.aborted, OperateAbortedIndexSnafu);
+
+        if self.creators.is_empty() {
+            return Ok(());
+        }
+
+        if let Err(update_err) = self.do_update(batch).await {
+            // clean up garbage if failed to update
+            if let Err(err) = self.do_cleanup().await {
+                if cfg!(any(test, feature = "test")) {
+                    panic!("Failed to clean up index creator, err: {err}",);
```

Comment:
We may need to use debug format to get the root cause
```suggestion
                    panic!("Failed to clean up index creator, err: {err:?}",);
```